### PR TITLE
Add --parallel N flag to vera-bench run

### PR DESCRIPTION
**Pull request by sunholo-voight-kampff**
## Summary

Adds a `--parallel N` flag to `vera-bench run` that dispatches problems to a `ThreadPoolExecutor` with `N` workers. Each worker is I/O-bound on its LLM call + subprocess `check`/`run`, so the GIL is not a bottleneck.

**Use case:** slow models like Kimi K2.5 averaged ~50s/problem sequentially across the 60-problem sweep (~50 min total). With `--parallel 10` that should drop to ~5 min, which makes release-time re-evals practical.

**Default `parallel=1` preserves the existing sequential path with no behavior change.**

## Scope note

This was originally bundled into [PR #70](https://github.com/aallan/vera-bench/pull/70) (the AILANG baseline). It snuck in during release-time benchmarking and was always intended to be a separate PR — concurrent sweep is independent of AILANG support and deserves its own review. Reverted out of #70 ([revert commit](https://github.com/aallan/vera-bench/pull/70/commits)) and surfacing here.

## Implementation

- `ThreadPoolExecutor` with `max_workers=parallel`, futures collected via `as_completed`
- `threading.Lock` around the JSONL append so concurrent writes don't interleave. Lines are still self-contained (carry `problem_id`) so completion-order writes are fine for downstream consumers
- Workers share `work_dir`; per-problem temp files are uniquified by `problem_id` (existing behaviour, unchanged)
- Exception in any worker is caught, logged red, and the sweep continues — one crashed problem doesn't kill the rest

Smoke-tested with `claude-haiku-4-5 --tier 1 --parallel 4`: 10/10 problems, no duplicates, 100%/100% run_correct.

## Tests

`TestRunBenchmarkParallel` in `tests/test_runner.py` adds 7 cases covering:

- `parallel=1` does NOT use `ThreadPoolExecutor` (patched to raise on use)
- `parallel>1` runs every problem and collects every result
- One worker raising doesn't abort the sweep
- `write_lock` serialises JSONL writes (20 problems × 8 workers, every line parseable JSON)
- `output_path=None` is a valid code path
- Click accepts `--parallel N` and defaults to 1

All 7 pass locally; `ruff check .` / `ruff format --check .` / `ruff check --select S vera_bench/` all clean. Local coverage on `vera_bench/runner.py`: 83% (CI lifts further when vera-binary-dependent paths run).

## Test plan

- [ ] CI green
- [ ] Code review

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * `run` gains a `--parallel` option (default 1) to run problems concurrently; progress shows chosen parallelism, worker errors are recorded per problem without aborting the run, and metadata versions are propagated. `output_path=None` returns in-memory results without writing files.

* **Tests**
  * Added tests covering sequential and parallel execution, concurrency-safe output integrity, error isolation and reporting, CLI validation for `--parallel`, and progress updates for every problem.

[Review Change Stack](https://app.coderabbit.ai/change-stack/aallan/vera-bench/pull/73?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

**Comment by sunholo-voight-kampff** [comment]
@coderabbitai review

(Re-triggering — the initial review at 13:32Z was rate-limited.)

File: tests/test_runner.py [hunk 1]
```diff
@@ -1508,3 +1508,217 @@ def test_writes_jsonl(self, tmp_path):
...
+                "claude-haiku-4-5-20251001",
+                "--parallel",
+                "4",
+                "--problem",
+                "VB-T1-001",
+            ],
+        )
+        assert "no such option" not in (result.output or "").lower()
+        assert "invalid" not in (result.output or "").lower()
+
```
[the remark this answers is not in the document]
Comment (sunholo-voight-kampff) [reply tests/test_runner.py:1970]: @coderabbitai Addressed in [bfbfae2](https://github.com/aallan/vera-bench/pull/73/commits/bfbfae2). Applied your suggestion verbatim: replaced `assert "invalid" not in (result.output or "").lower()` with `assert result.exit_code != 2`.

Also added two new tests pinning the new `IntRange(min=1)` behaviour from the cli.py finding so 0/negative are guaranteed to fail at parse time going forward:
- `test_run_command_rejects_zero_parallel`: `--parallel 0` → exit 2
- `test_run_command_rejects_negative_parallel`: `--parallel -5` → exit 2

Closing out.

File: vera_bench/cli.py [hunk 2]
```diff
@@ -76,6 +76,17 @@ def validate(problems_dir: Path | None, solutions_dir: Path | None):
     is_flag=True,
     help="Keep temporary generated files",
 )
+@click.option(
+    "--parallel",
+    type=int,
+    default=1,
+    show_default=True,
```
[the remark this answers is not in the document]
Comment (sunholo-voight-kampff) [reply vera_bench/cli.py:83]: @coderabbitai Addressed in [bfbfae2](https://github.com/aallan/vera-bench/pull/73/commits/bfbfae2) — applied your `click.IntRange(min=1)` suggestion verbatim. Two new tests in `TestRunBenchmarkParallel` pin the parse-time rejection of 0 and negative values (both return Click's `exit_code == 2`).

Declined the `-> None` annotation on `def run(...)`: per @aallan's prior comment on PR #70 (in the commit description for [aa13f25](https://github.com/aallan/vera-bench/commit/aa13f25)), missing `-> None` applies to **all** Click handlers in cli.py as a pre-existing project-wide consistency item — not specific to this change. Annotating just `run` would break that consistency. Happy to revisit if @aallan wants a dedicated PR sweeping all four handlers.

Closing out.

File: vera_bench/runner.py [hunk 3]
```diff
@@ -864,20 +864,62 @@ def run_benchmark(
...
+                            for r in problem_results:
+                                f.write(r.to_jsonl() + "\n")
+
+                    progress.advance(task)
+        else:
+            import threading
+            from concurrent.futures import ThreadPoolExecutor, as_completed
+
+            write_lock = threading.Lock()
+
```
[the remark this answers is not in the document]
Comment (sunholo-voight-kampff) [reply vera_bench/runner.py]: @coderabbitai Addressed in [bfbfae2](https://github.com/aallan/vera-bench/pull/73/commits/bfbfae2). Sharp finding — you were exactly right: the `for fut in as_completed(...)` loop runs on the main thread, workers only run `_run_one` (LLM/subprocess work), so the JSONL appends were already serialised by the loop structure. The `write_lock` and `import threading` were noise pretending to be protection.

Applied your simplification verbatim. Also updated the docstring on `test_parallel_writes_are_serialised` so it credits the loop (not the lock) as the serialisation source — the test's assertion is still valid (every line is parseable JSON under load), but the *why* now matches reality. Added a comment in the source explaining where serialisation comes from, so a future reader doesn't re-add the lock thinking it was load-bearing.

Closing out.

File: tests/test_runner.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (aallan) [reply tests/test_runner.py:1970]: Confirmed resolved from maintainer side — `exit_code != 2` is the right semantic for Click parse failures, and the two new `IntRange(min=1)` pin tests (`test_run_command_rejects_zero_parallel` / `test_run_command_rejects_negative_parallel`) close the regression class properly. Nice work on the extra coverage.

@coderabbitai resolve

File: vera_bench/cli.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (aallan) [reply vera_bench/cli.py:83]: `IntRange(min=1)` portion confirmed resolved from maintainer side.

On the `-> None` decline: the asymmetry argument is the right call. The three sibling `@click.command` handlers in `cli.py` are similarly unannotated, so a single-handler annotation introduces noise. Deferred to a separate sweep that does all four.

@coderabbitai resolve

File: vera_bench/runner.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (aallan) [reply vera_bench/runner.py]: Confirmed resolved from maintainer side. Removing the lock + the explanatory comment about main-thread serialisation via `as_completed` is the right call — that comment is now load-bearing and will prevent re-introduction by a future contributor who didn't trace the concurrency model. Sharp execution on the diagnosis.

@coderabbitai resolve

**Review by aallan — CHANGES_REQUESTED** [review]
## Detailed review — `--parallel N`

Ran a four-agent parallel review (code-reviewer, pr-test-analyzer, silent-failure-hunter, comment-analyzer). Findings are deduped; convergence noted where multiple agents independently agreed.

### TL;DR

No critical merge-blockers — threading model is sound, JSONL serialisation is correct, CI clean across all 9 checks, all three CodeRabbit inline findings substantively addressed in `bfbfae2`. But the four agents converged on five **Important** issues worth a second iteration before merge, plus five negotiable suggestions. The standout is **I1** (worker crashes vanish from JSONL) — flagged by silent-failure-hunter and arguably the closest thing to a blocker, since it makes parallel sweeps silently under-report scope to downstream `vera-bench report`.

---

### Important Issues (5)

#### I1 — Worker crashes vanish from JSONL ★ **priority**

`vera_bench/runner.py:945-950`. When `fut.result()` raises, the handler logs `[red]Worker failed on {pid}: {exc}[/red]` to stdout and `continue`s — **no `ProblemResult` is written to `output_path`**. Downstream consequences:

- A 60-problem sweep with 2 worker crashes produces 58 JSONL lines.
- `vera-bench report` then computes pass/fail metrics over those 58 lines and reports "58/58 succeeded (100%)" — the 2 crashes are invisible to anyone reading results files later.
- The sequential path doesn't have this defect because crashes abort the whole sweep loudly with a traceback.

For a benchmark tool, silently changing the denominator on operator error is arguably worse than diagnostic loss. Fix:

```python
except Exception as exc:  # noqa: BLE001
    pid = futures[fut].get("id", "?")
    tb = traceback.format_exc()
    console.print(f"[red]Worker failed on {pid}: {exc!r}[/red]")
    console.print(f"[dim]{tb}[/dim]")
    crash_result = ProblemResult(
        problem_id=pid,
        model=getattr(client, "model", "unknown"),
        language=language,
        attempt=0,
        check_pass=False,
        run_correct=False,
        error_message=f"Worker crash: {exc!r}\n{tb}",
        timestamp=datetime.now(timezone.utc).isoformat(),
        # ... other required fields
    )
    all_results.append(crash_result)
    if output_path:
        with open(output_path, "a", encoding="utf-8") as f:
            f.write(crash_result.to_jsonl() + "\n")
    progress.advance(task)
    continue
```

This single change addresses I1 **and** I3 (traceback) at once.

#### I2 — Sequential/parallel error-handling asymmetry

Two agents converged. `runner.py:885-906` (sequential, no try/except) vs `runner.py:945-952` (parallel, swallows `Exception`). `vera-bench run --parallel 1` and `vera-bench run --parallel 2` have different fault semantics on the same input — a transient `NoneType` from a bad model response kills the whole sweep at `--parallel 1` but is logged-and-continued at `--parallel 2`.

Either align them (recommended — wrap the sequential body in the same try/except so a 4-hour sweep survives problem 47 of 60 regardless of parallelism), or document the asymmetry in the `run_benchmark` docstring alongside the existing "JSONL output ordering" note.

#### I3 — Lost traceback in worker exception handler

Same diagnostic-loss pattern as **C3 on PR #70**. `runner.py:947` reports `str(exc)` only. On a 1-hour parallel sweep with hundreds of problems, debugging `Worker failed on VB-T3-007: 'NoneType' object has no attribute 'foo'` with no file/line is painful.

Subsumed by the I1 fix above (capture `traceback.format_exc()` into the JSONL row's `error_message`).

#### I4 — `progress.advance(task)` on exception path is untested

`tests/test_runner.py:1591-1620` (`test_parallel_worker_exception_continues`) verifies the loop continues (3 of 4 results return), but does not pin that the progress bar advanced on the crashed problem. The current code does call `advance` before `continue` at `runner.py:951`, but a refactor moving it into an `else:` branch would silently regress the bar to hang at N-1/N with no test failure.

Fix: inject/patch `progress` (or pass a fake) and assert `advance.call_count == len(problems)` even with a worker exception.

#### I5 — Version propagation through `_run_one` closure is untested

`runner.py:922-936`. `bench_version` and `vera_version` are captured by closure into `_run_one`. If a future refactor drops them from the kwargs forwarded to `run_single_problem`, JSONL would silently get empty version strings. The existing mock at `test_runner.py:1546` uses `lambda problem, **kw: ...` which swallows the kwargs without inspection.

Fix: one test that calls with `bench_version="0.1.0", vera_version="0.0.103"` and asserts the mock saw them.

---

### Suggestions (5, negotiable)

- **S1 — Output-write block has no error handling.** `runner.py:949` — disk full or NFS hiccup mid-sweep raises `OSError` on the main thread, killing the whole sweep. Pre-existing in sequential too, but parallel makes the blast radius bigger (more in-flight uncommitted results). Consider try/except around the write with a fallback log.
- **S2 — Patch-target fragility.** `tests/test_runner.py:1538-1541` patches `concurrent.futures.ThreadPoolExecutor`. Works because the import is lazy in the parallel branch, but would silently fail if someone moves the import to module scope. Robustness alternative: assert `threading.current_thread() is threading.main_thread()` inside `_run_one` — asserts behaviour (no thread spawn) rather than implementation (no class lookup).
- **S3 — Anecdotal Kimi K2.5 performance figures.** `runner.py:877-879` and `cli.py:85` cite "~50s → ~5s" with no committed benchmark, and the model name will age out. Either commit a measurement or soften to generic "slow models".
- **S4 — Misleading POSIX-atomicity claim in test docstring.** `tests/test_runner.py:1623`'s "Python's GIL doesn't make file writes atomic — partial writes are observable" is wrong for the actual case here: short writes (<PIPE_BUF ~4096B) with `O_APPEND` are atomic on POSIX, and the production code serialises writes on the main thread anyway. The test still verifies useful properties; the docstring oversells.
- **S5 — 20×8 stress level is overkill post-lock-removal.** With writes now serialised on main thread, the test only catches the regression of moving writes back into workers — a single race at much lower scale would catch it. Lower N if test suite gets slow.

---

### Strengths

Substantial — calling out specifically because four agents converged on these:

- **`runner.py:919-922` "no write lock needed" comment** — load-bearing and exemplary (called out by three agents). The kind of comment that survives refactors and actively prevents a future contributor from re-introducing a redundant lock thinking it was protective. Best comment in the PR.
- **`cli.py:79-89` `click.IntRange(min=1)` with `show_default=True`** — closes the silent-fall-through-to-sequential bug class that plain `type=int` would have allowed. Right pin at the parse boundary.
- **`tests/test_runner.py:1518-1564` `test_parallel_one_uses_sequential_path`** — clever negative-space test. Patching `ThreadPoolExecutor` with `side_effect=AssertionError("must not be used in sequential path")` proves the *absence* of an import path, which is genuinely harder than asserting presence.
- **`runner.py:946` `# noqa: BLE001` correctly catches `Exception` not `BaseException`** — Ctrl-C / `SystemExit` still propagate and cancel the sweep, which is exactly what you want. Two agents called this out as the right narrow choice.
- **Set-equality assertions** at `tests/test_runner.py:1581`, `1611` correctly accommodate completion-order nondeterminism. Avoids the trap of list-equality assertions that would be flaky under parallel scheduling.
- **CLI tests assert `exit_code != 2`** rather than message-string matching — resilient to Click phrasing changes, the right semantic boundary for parse-time validation.
- **Inline closure `_run_one`** at `runner.py:922-936` is the right abstraction level — captures 9 invariants locally without polluting module namespace. A module-level `_dispatch_one` would add boilerplate for no real win.
- **`run_benchmark` docstring** at `runner.py:872-883` honestly documents completion-order JSONL ordering with mitigation (`problem_id` is self-contained, consumers can sort). Setting expectations correctly upfront.
- **Review-iteration discipline** — applied 3/3 CodeRabbit inline findings verbatim with extra coverage (the two new `IntRange` pin tests went beyond what CR asked for), and declined the `-> None` annotation with documented rationale rather than silently ignoring. Exactly the response pattern that makes review productive.

---

### Recommended action

This is close to merge-ready. I'd ask for one more iteration on **I1-I5**:

1. **I1 + I3 in one change** — synthesise a `ProblemResult` on worker crash with `traceback.format_exc()` in `error_message`. The patch sketch above is ~10 lines and lands both fixes.
2. **I2** — wrap the sequential body in the same try/except so semantics match, OR document the asymmetry. Either is defensible.
3. **I4 + I5** — small test additions in the existing `TestRunBenchmarkParallel` class.
4. **S1-S5 are negotiable** — your call. None are blocking.

After those land I'd approve and merge. Thanks for the careful work on this — the CR-iteration discipline and the load-bearing concurrency comment are both first-rate.

— Reviewed with a four-agent parallel pass (general code review, test coverage, silent-failure hunting, comment accuracy). Convergence noted above where multiple agents flagged the same issue.

**Comment by sunholo-voight-kampff** [comment]
@aallan — addressed in [fa8d1ce](https://github.com/aallan/vera-bench/pull/73/commits/fa8d1ce). Going item-by-item against your four-agent review:

### Important (all in)

- [x] **I1 + I3 (priority)** — Worker crashes now synthesise a visible `ProblemResult` row with `check_pass=False`, `run_correct=False`, and `traceback.format_exc()` embedded in `error_message`. Extracted into a `_crash_result(problem, exc, tb)` helper and a `_record(problem_results)` helper (both nested in `run_benchmark` to capture the closure variables cleanly) so successes and crashes hit the same persistence path. Your diff sketch caught both findings in one change — landed essentially verbatim, with the helper extracted so the sequential path can reuse it.
- [x] **I2 (sequential / parallel symmetry)** — Took the alignment option (recommended). Sequential `run_single_problem` call is now wrapped in the same `try/except`, routing crashes through `_crash_result` + `_record`. A 4-hour sweep now survives problem 47 of 60 regardless of `--parallel N`.
- [x] **I4 (`progress.advance` on exception path)** — `test_progress_advances_on_crash_path` patches `Progress` and asserts `advance.call_count == len(problems)` even when one problem raises, in **both** sequential and parallel paths. Refactor that moves `advance` into an `else:` branch would now fail this test.
- [x] **I5 (version propagation through closure)** — `test_bench_and_vera_version_propagate_to_workers` captures the kwargs `run_single_problem` actually receives under `parallel=3` and asserts both `bench_version` and `vera_version` came through. Catches a future refactor dropping them.

### Suggestions

- [x] **S2 (patch-target fragility)** — Replaced the `ThreadPoolExecutor` `side_effect=AssertionError(...)` patch with a thread-identity assertion: under `parallel=1`, every `run_single_problem` call must run on `threading.main_thread()`. Robust to a future refactor hoisting the import to module scope. Added a counterpoint test (`test_parallel_two_actually_spawns_worker_threads`) confirming `parallel>1` actually spawns workers, since the negative-only test could otherwise pass if the parallel path was silently broken too.
- [x] **S4 (incorrect POSIX-atomicity claim)** — Fixed. The old docstring claimed "Python's GIL doesn't make file writes atomic — partial writes are observable", which is wrong: short `O_APPEND` writes < PIPE_BUF are atomic on POSIX. Replaced with an honest description: the test proves the main-thread `as_completed` loop is the serialisation source, not POSIX guarantees we don't depend on.

- [ ] **S1 (no error handling on output write)** — Deferred. Pre-existing on the sequential path too, and a sensible operator response (resume from JSONL) doesn't exist yet, so wrapping in try/except just to log-and-continue past a disk-full would lose data without a recovery path. Happy to revisit alongside a resume-from-jsonl feature.
- [ ] **S3 (anecdotal Kimi K2.5 figures)** — Kept as-is. They're motivating context in the docstring, not a load-bearing claim downstream code relies on. If you'd prefer them softened or removed I'll do that.
- [ ] **S5 (20×8 stress overkill)** — Kept. Test runtime is sub-second and the larger scale catches more refactor failure modes than a single race would.

### Updated existing test for new behavior

`test_parallel_worker_exception_continues` previously asserted `len(results) == 3` (crashed problem vanished) and 3 JSONL lines. Now asserts `len(results) == 4` and verifies the crash row carries `Worker crash:`, the original exception's repr, and a `Traceback` in `error_message`. Plus a new `test_sequential_worker_exception_also_continues` pinning the I2 symmetry.

13 `TestRunBenchmarkParallel` cases pass locally; `ruff check .` / `ruff format --check .` / `ruff check --select S vera_bench/` all clean. Pushing now; CI will run momentarily.

Thanks again for the substantive review — the I1 finding (silent denominator change) was a real one I would not have caught, and the four-agent convergence on the strengths section was good signal too.

File: tests/test_runner.py [hunk 4]
```diff
@@ -1508,3 +1508,461 @@ def test_writes_jsonl(self, tmp_path):
...
+        scope (legitimate change) would still pass this test."""
+        import threading
+
+        from vera_bench.runner import run_benchmark
+
+        calling_threads: list[threading.Thread] = []
+
+        def _record_thread(problem, **kw):
+            calling_threads.append(threading.current_thread())
+            return [self._result(problem["id"])]
```
[the remark this answers is not in the document]
Comment (sunholo-voight-kampff) [reply tests/test_runner.py]: @coderabbitai Addressed in [7a161f8](https://github.com/aallan/vera-bench/pull/73/commits/7a161f8). All six inner helpers (`_record_thread` ×2, `_side_effect` ×3, `_capture`) now annotated:

```python
def _xyz(
    problem: dict[str, object], **kw: object
) -> list[ProblemResult]
```

`ProblemResult` was already imported at module scope. Closing out.

File: tests/test_runner.py [hunk 5]
```diff
@@ -1508,3 +1508,461 @@ def test_writes_jsonl(self, tmp_path):
...
+        # ALL 4 problems produce a result row — 3 successes + 1 crash.
+        assert len(results) == 4
+        ids = {r.problem_id for r in results}
+        assert ids == {"VB-X-0", "VB-X-1", "VB-X-2", "VB-X-3"}
+        # JSONL has 4 lines (the crash row IS written so report doesn't
+        # silently shrink the denominator).
+        lines = output.read_text().strip().split("\n")
+        assert len(lines) == 4
+        # The crash row carries diagnostic detail in `error_message`.
+        crash_row = next(json.loads(ln) for ln in lines if "Worker crash" in ln)
```
[the remark this answers is not in the document]
Comment (sunholo-voight-kampff) [reply tests/test_runner.py]: @coderabbitai Addressed in [7a161f8](https://github.com/aallan/vera-bench/pull/73/commits/7a161f8). Applied your diff verbatim — both `test_parallel_worker_exception_continues` and `test_sequential_worker_exception_also_continues` now use:

```python
rows = [json.loads(ln) for ln in lines]
crash_row = next(row for row in rows if row["problem_id"] == "<id>")
```

The message-content assertions (`"simulated worker crash"`, `"RuntimeError"`, `"Traceback"`) remain — they're now explicitly testing the message-content contract rather than depending on it implicitly through the selector. Behaviour pinned without coupling to wording. Closing out.

**Review by aallan — APPROVED** [review]
## Approving — everything in scope is delivered

All five Important issues from the 2026-05-22T20:44 review are resolved, two of five suggestions taken with documented rationale on the rest, and both follow-up CodeRabbit findings on the response commit have already been addressed and confirmed by CR. CI is green across all 9 checks on `7a161f84`.

### Verification

| Ask | Delivered | Location |
|---|---|---|
| **I1 + I3** — worker crashes vanish from JSONL + lost traceback | Unified `_crash_result()` synthesises a `ProblemResult` with `traceback.format_exc()` in `error_message`; new `_record()` helper unifies in-memory + JSONL persistence between success and crash paths | `runner.py:894-924` |
| **I2** — sequential / parallel error-handling asymmetry | Both paths now wrap `run_single_problem` in identical `try/except` and route crashes through `_crash_result + _record`. Same fault semantics regardless of `--parallel` N | `runner.py:927-1001` |
| **I4** — `progress.advance(task)` untested on exception path | New `test_progress_advances_on_crash_path` patches `Progress`, runs both `parallel=2` and `parallel=1` back-to-back, asserts `advance.call_count == 4` (3 successes + 1 crash) on each path | `tests/test_runner.py:1814-1868` |
| **I5** — version propagation through closure untested | New `test_bench_and_vera_version_propagate_to_workers` captures the kwargs `run_single_problem` actually receives via a `_capture` closure under `parallel=3` and asserts both versions arrive on *every* call | `tests/test_runner.py:1870-1908` |
| **S2** — patch-target fragility | Replaced `ThreadPoolExecutor` patch with `threading.current_thread() is threading.main_thread()` assertion (behaviour over implementation); added counterpoint `test_parallel_two_actually_spawns_worker_threads` | `tests/test_runner.py:1543-1625` |
| **S4** — misleading POSIX-atomicity docstring | Rewritten to honest "main-thread `as_completed` serialisation" explanation | `tests/test_runner.py:1749-1791` |

Deferred suggestions (S1, S3, S5) all have documented rationale in the commit message — file-write error handling is a pre-existing cross-cutting concern, the Kimi K2.5 anecdote is motivating context, and the 20×8 stress test runs sub-second so the larger scale is free.

### Worth calling out

The response on **I1 + I3** went structurally tighter than my review prompted. The minimal fix would have been "add `traceback.format_exc()` to the existing log line." Instead you refactored both paths through `_crash_result()` + `_record()` so successes and crashes hit the same persistence machinery — that abstraction makes "did this crash get recorded?" a property of the helper rather than a branching obligation at every call site. Future contributors who add a new branch (e.g., a timeout path) inherit the persistence semantics for free.

The **I4 test design** is also tighter than the brief: I asked for an assertion on the parallel path; you patched `Progress`, exercised *both* `parallel=2` and `parallel=1` back-to-back with a reset between, and pinned `advance.call_count == 4` on both. So I2's symmetric-fault-semantics design gets defended by I4's test on both paths in one go.

The **S2 thread-identity** replacement is the right kind of trade — implementation tests (patching imports) become brittle the moment someone moves an import; behaviour tests (assert this code runs on the main thread) survive that refactor cleanly. The counterpoint test ensures the inversion still holds.

And the **two follow-up CodeRabbit findings** on the response commit (type hints on six inner helpers, structural row selection by `problem_id` instead of message substring) were addressed in `7a161f84` within seven minutes of CR posting them, with CR confirming both. That iteration discipline keeps review productive.

### Approved

CI green, every priority ask delivered, deferred items explicitly tracked. Ready to merge.
